[docs] Batch 25 - MVD functions (#14)
edgar2020/druid: docs/querying/sql-functions.md

* [Window functions](sql-window-functions.md)

The examples on this page use the following example datasources:
* `flight-carriers` using `FlightCarrierOnTime (1 month)` included with Druid
* `kttm` using `KoalasToTheMax one day` included with Druid

> * `mvd_example` using the SQL-based ingestion example for [multi-value dimensions](multi-value-dimensions.md#sql-based-ingestion)

Reviewer: @ektravel included the SQL here with her PR. should standardize either include SQL or just link

<details><summary>Example</summary>

The following example appends the string `label` to the multi-value string `tags` from `mvd_example`:

```sql
SELECT MV_APPEND("tags", "label") AS append
FROM "mvd_example"
LIMIT 1
```

Returns the following:

| `append` |
| -- |
| `["t1","t2","t3","row1"]` |

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_CONCAT

Concatenates two arrays.

* **Syntax:** `MV_CONCAT(arr1, arr2)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example concatenates `tags` from `mvd_example` to itself:

```sql
SELECT MV_CONCAT("tags", "tags") AS cat
FROM "mvd_example"
LIMIT 1
```

Returns the following:

| `cat` |
| -- |
| `["t1","t2","t3","t1","t2","t3"]` |

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_CONTAINS

Returns true if the expression is in the array, false otherwise.

* **Syntax:** `MV_CONTAINS(arr, expr)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example checks whether the string `t3` is located within `tags` from `mvd_example`:

```sql
SELECT MV_CONTAINS("tags", 't3') AS contained
FROM "mvd_example"

> LIMIT 1

Reviewer: The following query might be a bit more illustrative since it shows the row & the value and includes both true & false:

```
SELECT tags, MV_CONTAINS("tags", 't3') AS contained
FROM "mvd_example"
```

<details><summary>Example</summary>

The following example returns the length of the `tags` multi-value strings from `mvd_example`:

```sql
SELECT MV_LENGTH("tags") AS len
FROM "mvd_example"
LIMIT 1
```

Returns the following:

| `len` |
| -- |
| `3` |

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_OFFSET

Returns the array element at the given zero-based index.

* **Syntax:** `MV_OFFSET(arr, long)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example returns the element at the third position of `tags` in `mvd_example`:

```sql
SELECT MV_OFFSET("tags", 2) AS elem
FROM "mvd_example"
LIMIT 1
```

Returns the following:

| `elem` |
| -- |
| `t3` |

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_OFFSET_OF

Returns the zero-based index of the first occurrence of a given expression in the array.

* **Syntax:** `MV_OFFSET_OF(arr, expr)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example returns the zero-based index of the string `t3` from `tags` in `mvd_example`:

```sql
SELECT MV_OFFSET_OF("tags", 't3') AS index
FROM "mvd_example"

> LIMIT 1

Reviewer: IMO the example is more illustrative without the limit. In general, I don't feel like the limit is doing a whole lot in this data source that has so few rows.

Additionally, recommend including the "tags" dimension in the result for offset & ordinal seems more illustrative.

* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example returns the overlap of `tags` and itself from `mvd_example`:

```sql
SELECT MV_OVERLAP("tags", "tags") AS overlap
FROM "mvd_example"
LIMIT 1
```

Returns the following:

| `overlap` |
| -- |
| `true` |

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_PREPEND

Adds the expression to the beginning of the array.

* **Syntax:** `MV_PREPEND(expr, arr)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example prepends the string dimension `label` to the multi-value string dimension `tags` from `mvd_example`:

```sql
SELECT MV_PREPEND("label", "tags") AS prepend
FROM "mvd_example"
LIMIT 1
```

Returns the following:

| `prepend` |
| -- |
| `["row1","t1","t2","t3"]` |

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_SLICE

Returns a slice of the array from the zero-based start and end indexes.

* **Syntax:** `MV_SLICE(arr, start, end)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example returns the second value of the `tags` column from `mvd_example`:

```sql

> SELECT MV_SLICE(tags, 1, 2) AS slice

Reviewer: suggest including "tags", bigger "slice", and removing limit

```
SELECT "tags", MV_SLICE(tags, 1, 3) AS slice
FROM "mvd_example"
```

Adds the expression to the end of the array.

* **Syntax:** `MV_APPEND(arr1, expr)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example appends the string `label` to the multi-value string `tags` from `mvd-example`:

```sql
SELECT MV_APPEND("tags", "label") AS append
FROM "mvd-example"
LIMIT 1
```

Returns the following:

| `append` |
| -- |
| `["t1","t2","t3","row1"]` |

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_CONCAT

Concatenates two arrays.

* **Syntax:** `MV_CONCAT(arr1, arr2)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example concatenates `tags` from `mvd-example` to itself:

```sql
SELECT MV_CONCAT("tags", "tags") AS cat
FROM "mvd-example"
LIMIT 1
```

Returns the following:

| `cat` |
| -- |
| `["t1","t2","t3","t1","t2","t3"]` |

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_CONTAINS

Returns true if the expression is in the array, false otherwise.

* **Syntax:** `MV_CONTAINS(arr, expr)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

> The following example checks whether the string `t3` is located within `tags` from `mvd-example`:

Reviewer: ```suggestion
The following example checks if the string `t3` exists within `tags` from `mvd-example`:
```
nit

```sql
SELECT "tags", MV_CONTAINS("tags", 't3') AS contained
FROM "mvd-example"
```

Returns the following:

|`tags`|`contained`|
|------|-----------|
|`["t1","t2","t3"]`|`true`|
|`["t3","t4","t5"]`|`true`|
|`["t5","t6","t7"]`|`false`|
|`null`|`false`|

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_FILTER_NONE

> Filters a multi-value expression to include no values contained in the array.

Reviewer: ```suggestion
Filters a multi-value expression to exclude values from an array.
```
nit

* **Syntax:** `MV_FILTER_NONE(expr, arr)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example filters `tags` from `mvd-example` to remove values `t1` or `t3`, if present:

```sql

> SELECT MV_FILTER_NONE("tags", ARRAY['t1', 't3']) AS nofilt

Reviewer: ```suggestion
SELECT MV_FILTER_NONE("tags", ARRAY['t1', 't3']) AS exclude
```

FROM "mvd-example"
LIMIT 3
```

Returns the following:

> | `nofilt` |

Reviewer: ```suggestion
| `exclude` |
```

FROM "mvd-example"
```

Returns the following:

|`tags`|`elem`|
|------|------|
|`["t1","t2","t3"]`|`t3`|
|`["t3","t4","t5"]`|`t5`|
|`["t5","t6","t7"]`|`t7`|
|`null`|`null`|

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_ORDINAL_OF

Returns the one-based index of the first occurrence of a given expression.

* **Syntax:** `MV_ORDINAL_OF(arr, expr)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example returns `tags` and the one-based index of the string `t3` from `tags` in `mvd-example`:

```sql
SELECT "tags", MV_ORDINAL_OF("tags", 't3') AS index
FROM "mvd-example"
```

Returns the following:

|`tags`|`index`|
|------|-------|
|`["t1","t2","t3"]`|`3`|
|`["t3","t4","t5"]`|`1`|
|`["t5","t6","t7"]`|`null`|
|`null`|`null`|

</details>

[Learn more](sql-multivalue-string-functions.md)

## MV_OVERLAP

Returns true if the two arrays have any elements in common, false otherwise.

* **Syntax:** `MV_OVERLAP(arr1, arr2)`
* **Function type:** Multi-value string

<details><summary>Example</summary>

The following example returns the overlap of `tags` and itself from `mvd-example`:

```sql

> SELECT MV_OVERLAP("tags", "tags") AS overlap

Reviewer: Kind of a weird example since an MVD will always have overlap with itself.
